torch/package/_vinegar_check.py

```diff
@@ -0,0 +1,347 @@
+        assert len(self.stack) == 0
+        return ret
+
+    def pop_mark(self):
+        items = self.stack
+        self.stack = self.metastack.pop()
+        self.append = self.stack.append
+        return items
+
+    def load_simple_push(self, arg, pos):
+        self.append(arg)
+
+    def load_empty_tuple(self, arg, pos):
+        self.append(TrackedTuple((), pos))
+
+    def load_empty_dict(self, arg, pos):
+        self.append(TrackedDict(pos))
+
+    def load_empty_list(self, arg, pos):
+        self.append(TrackedList(pos))
+
+    def load_tuple(self, arg, pos):
+        items = self.pop_mark()
+        self.append(TrackedTuple(items, pos))
+
+    def load_tuple1(self, arg, pos):
+        self.stack[-1] = TrackedTuple(self.stack[-1:], pos)
+
+    def load_tuple2(self, arg, pos):
+        self.stack[-2:] = [TrackedTuple(self.stack[-2:], pos)]
+
+    def load_tuple3(self, arg, pos):
+        self.stack[-3:] = [TrackedTuple(self.stack[-3:], pos)]
+
+    def load_none(self, arg, pos):
+        self.append(None)
+
+    def load_newtrue(self, arg, pos):
+        self.append(True)
+
+    def load_newfalse(self, arg, pos):
+        self.append(False)
+
+    def load_binpersid(self, arg, pos):
+        pass
+
+    def load_memo_put(self, arg, pos):
+        self.memo[arg] = self.stack[-1]
+
+    def load_memo_get(self, arg, pos):
+        self.append(self.memo[arg])
+
+    def load_global(self, arg, pos):
+        # GLOBAL is returned a string, not class/fn
+        self.append(arg)
+
+    def load_reduce(self, arg, pos):
+        args = self.stack.pop()
+        func = self.stack.pop()
+        if func == "collections OrderedDict":
+            # hack to make this work
+            self.append(TrackedDict(pos))
+        else:
+            ret = TrackedObj(func, pos)
+            children = {}
+            for i, arg in enumerate(args.children.values()):
+                arg_name = f"reduce_arg_{i}"
+                children[arg_name] = arg
+                if isinstance(arg, PickleObjTree):
+                    arg.attr_name = arg_name
+                    arg.parent = ret
+            ret.children = children
+            self.append(ret)
+
+    def load_newobj(self, arg, pos):
+        args = self.stack.pop()
+        klass = self.stack.pop()
+        obj = TrackedObj(klass, pos)
+        self.append(obj)
+
+    def load_build(self, arg, pos):
+        state = self.stack.pop()
+        obj = self.stack[-1]
+        if isinstance(state, TrackedDict):
+            # this is a regular __dict__ update.
+            assert len(obj.children) == 0
+            obj.children.update(state)
+
+            # Annotate any objs with parent and attr information for better unwinding.
+            for k, v in state.items():
+                if not isinstance(v, PickleObjTree):
+                    continue
+                v.parent = obj
+                v.attr_name = k
+        elif isinstance(state, TrackedTuple):
```

Comment:
@suo the only logic I really modified in the pickle analyzer was adding logic to handle objects that have __slots__